File: android/PracticeAndroid/app/src/main/java/com/example/practiceandroid/MainActivity.java
package com.example.practiceandroid;

import android.os.Bundle;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

public class MainActivity extends AppCompatActivity {
    FragmentManager fm; // 프래그먼트 매니저
    FragmentTransaction ft; // 프래그먼트 트랜잭션

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        fm = getSupportFragmentManager(); // 프래그먼트 매니저 참조 객체 생성
        ft = fm.beginTransaction(); // 프래그먼트 매니저 시작
    }

    public void showFragment(View view) {
        ft = fm.beginTransaction();
        if (view == findViewById(R.id.button1)) {
            FragmentFirst fr1 = new FragmentFirst();
            ft.replace(R.id.fragment_container, fr1); // 프래그먼트 교체 메서드 replace(교체될 프래그먼트가 보여질 최상위 루트뷰, 교체될 프래그먼트)
            ft.commit(); // 변경 내용 적용을 위한 프래그먼트 커밋
        } else {
            FragmentSecond fr2 = new FragmentSecond();
            ft.replace(R.id.fragment_container, fr2); // 프래그먼트 교체 메서드 replace(교체될 프래그먼트가 보여질 최상위 루트뷰, 교체될 프래그먼트)
            ft.commit();
        }
    }
}

File: android/QRCodeReader/app/src/main/java/com/logmind/qrcodereader/MainActivity.kt
package com.logmind.qrcodereader

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.core.CameraSelector
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.Preview
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.google.common.util.concurrent.ListenableFuture
import com.logmind.qrcodereader.databinding.ActivityMainBinding
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var cameraProviderFuture: ListenableFuture<ProcessCameraProvider>
    private val PERMISSIONS_REQUEST_CODE = 1
    private val PERMISSIONS_REQUIRED = arrayOf(Manifest.permission.CAMERA)
    private var isDetected = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        if (!hasPermissions(this)) {
            requestPermissions(PERMISSIONS_REQUIRED, PERMISSIONS_REQUEST_CODE)
        } else {
            startCamera()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == PERMISSIONS_REQUEST_CODE) {
            if (PackageManager.PERMISSION_GRANTED == grantResults.firstOrNull()) {
                Toast.makeText(this@MainActivity, "권한 요청이 승인되었습니다.", Toast.LENGTH_LONG).show()
                startCamera()
            } else {
                Toast.makeText(this@MainActivity, "권한 요청이 거부되었습니다.", Toast.LENGTH_LONG).show()
                finish()
            }
        }
    }

    fun hasPermissions(context: Context) = PERMISSIONS_REQUIRED.all {
        ContextCompat.checkSelfPermission(context, it) ==
                PackageManager.PERMISSION_GRANTED
    }

    fun startCamera() {
        cameraProviderFuture = ProcessCameraProvider.getInstance(this)
        cameraProviderFuture.addListener(Runnable {
            val cameraProvider = cameraProviderFuture.get()

            val preview = getPreview()
            val imageAnalysis = getImageAnalysis()
            val cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA

            cameraProvider.bindToLifecycle(this, cameraSelector, preview, imageAnalysis)
        }, ContextCompat.getMainExecutor(this))
    }

    fun getPreview(): Preview {
        val preview: Preview = Preview.Builder().build()
        /**
         * SurfaceProvider는 Preview에 Surface를 제공해주는 인터페이스이다.
         * 화면에 보여지는 픽셀들이 모여 있는 개체가 Surface라고 생각하자.
         */
        preview.setSurfaceProvider(binding.barcodePreview.surfaceProvider)

        return preview
    }

    override fun onResume() {
        super.onResume()
        isDetected = false
    }

    fun getImageAnalysis(): ImageAnalysis {
        val cameraExecutor: ExecutorService = Executors.newSingleThreadExecutor()
        val imageAnalysis = ImageAnalysis
            .Builder()
            .build()

        imageAnalysis.setAnalyzer(
            cameraExecutor,
            QRCodeAnalyzer(object : OnDetectListener {
                override fun onDetect(msg: String) {
                    if (!isDetected) {
                        isDetected = true

                        val intent = Intent(this@MainActivity, ResultActivity::class.java)
                        intent.putExtra("msg", msg)
                        startActivity(intent)
                    }
                }
            })
        )
        return imageAnalysis
    }
}